File: problems/3186_maximum_total_damage_with_spell_casting/3186_maximum_total_damage_with_spell_casting.java
public class Solution {
    public long maximumTotalDamage(int[] power) {
        Map<Integer, Integer> cnt = new HashMap<>();
        for (int x : power)
            cnt.put(x, cnt.getOrDefault(x, 0) + 1);
        List<Integer> nums = new ArrayList<>(cnt.keySet());
        Collections.sort(nums);
        int n = nums.size();
        long[] dp = new long[n + 1];
        long[] total = new long[n];
        for (int i = 0; i < n; i++)
            total[i] = (long) nums.get(i) * cnt.get(nums.get(i));
        for (int i = 0; i < n; i++) {
            int j = i - 1;
            while (j >= 0 && nums.get(j) >= nums.get(i) - 2)
                j--;
            long pick = total[i] + (j >= 0 ? dp[j + 1] : 0);
            dp[i + 1] = Math.max(dp[i], pick);
        }
        return dp[n];
    }
}

File: problems/3186_maximum_total_damage_with_spell_casting/3186_maximum_total_damage_with_spell_casting.py
class Solution:
    def maximumTotalDamage(self, power: List[int]) -> int:
        cnt = Counter(power)
        nums = sorted(cnt)
        n = len(nums)
        dp = [0] * (n + 1)

        # Presum for each damage
        total = [nums[i] * cnt[nums[i]] for i in range(n)]
        
        for i in range(n):
            j = i - 1
            while j >= 0 and nums[j] >= nums[i] - 2:
                j -= 1
            pick = total[i] + (dp[j + 1] if j >= 0 else 0)
            dp[i + 1] = max(dp[i], pick)
        return dp[n]